### PR TITLE
Instantiate types in record literals as necessary

### LINKED ISSUE #1110: Inferred types of record fields should not be polymorphic
``` purescript
module Bug where

-- from purescript-maps
foreign import data StrMap :: * -> *
foreign import empty "" :: forall a. StrMap a
foreign import singleton "" :: forall a. String -> a -> StrMap a

-- from my code
type R = {f :: StrMap Number}

funA :: forall m . (Monad m) => m R
--funA = return {f:singleton "" 0} -- GOOD
funA = return {f:empty} -- BUG with this message:
```

``` text
Error in module Bug:
Error in value declaration funA:
Error checking that type
  u1499 u1498
subsumes type
  m1497 { f :: Bug.StrMap Prim.Number }
Error at bug2.purs line 17, column 1 - line 17, column 23:
  Cannot unify type
    a1501
  with type
    Prim.Number
See https://github.com/purescript/purescript/wiki/Error-Code-ErrorUnifyingTypes for more information, or to contribute content related to this error.
```

### PR DESCRIPTION
**Pull request by paf31**
Fixes #1110 and #2312. Replaces #2367.

I'm not really very happy with this fix, but I've been stuck thinking about this for a long time now, and it's been difficult to fix this and keep Halogen compiling as-is. So this is a compromise, and I've added a note in a comment to explain the fix.

### CODE REVIEW

File: src/Language/PureScript/TypeChecker/Types.hs [hunk 1]
```diff
@@ -315,10 +316,25 @@ infer' (Literal (ArrayLiteral vals)) = do
   return $ TypedValue True (Literal (ArrayLiteral ts')) (TypeApp tyArray els)
 infer' (Literal (ObjectLiteral ps)) = do
   ensureNoDuplicateProperties ps
-  ts <- traverse (infer . snd) ps
-  let fields = zipWith (\name (TypedValue _ _ t) -> (Label name, t)) (map fst ps) ts
-      ty = TypeApp tyRecord $ rowFromList (fields, REmpty)
-  return $ TypedValue True (Literal (ObjectLiteral (zip (map fst ps) ts))) ty
+  -- We make a special case for Vars in record labels, since there are the
```
Comment (michaelficarra) [inline src/Language/PureScript/TypeChecker/Types.hs]: I think you meant "these".

Comment (paf31) [inline src/Language/PureScript/TypeChecker/Types.hs]: Thanks!

**Comment by paf31** [comment]
@LiamGoodacre Would you mind reviewing this one, please?

**Comment by LiamGoodacre** [comment]
Will do, I'll have a look through this tonight :)

**Comment by LiamGoodacre** [comment]
What would you expect to happen in this code:
```purescript
type Z t = forall x. t x -> (forall a. t a) -> t x
class C t where c :: Z t
instance cA :: C Array where c x _ = x
test2 :: forall m. Monad m => m { ccc :: Z Array }
test2 = pure { ccc: c }
```

**Comment by paf31** [comment]
I'm not sure, but I could try it. Why do you ask about that snippet specifically?

**Comment by LiamGoodacre** [comment]
Sorry, I forgot to add an explanation.  I tried added the above code to your test file.  The idea was to also test a polymorphic var with a constraint and also a quantified argument.  I probably shouldn't have lumped them together, but anyway: the test failed with an escaped skolem.  I would have expected it to succeed, so I was wondering if you thought the same.

**Comment by paf31** [comment]
Constrained functions in record literals are still a bit of a pain. Does it work if you annotate the property?

**Comment by LiamGoodacre** [comment]
With an annotation it still fails in the same way:
```purescript
test2 :: forall m. Monad m => m { ccc :: Z Array }
test2 = pure { ccc: (c :: Z Array) }
```

```
  1) Passing examples '1110.purs' should compile and run without error
       Error found:
       in module Main
       at /Users/liamgoodacre/git/purescript/examples/passing/1110.purs line 25, column 1 - line 25, column 51

         A type variable has escaped its scope.

       in the expression { ccc: c cA
                         }
       in value declaration test2
```

**Comment by paf31** [comment]
Oh, then that's a bug. Thanks.

**Comment by paf31** [comment]
I think it's easily fixed, thankfully. I'll add a test for this.

**Comment by paf31** [comment]
@LiamGoodacre How's this?

**Comment by paf31** [comment]
The whole package set continues to compile with this change, so it seems safe enough.

**Comment by LiamGoodacre** [comment]
This looks good :) - although I'm not an expert on this part of the code.

**Pull request merged**